linkerd/http-retry/src/lib.rs

```diff
@@ -301,31 +311,29 @@ where
             && is_inner_eos
     }
 
-    fn size_hint(&self) -> http_body::SizeHint {
-        let mut hint = http_body::SizeHint::default();
-        if let Some(ref state) = self.state {
-            let rem = state.buf.remaining() as u64;
-
-            // Have we read the entire body? If so, the size is exactly the size
-            // of the buffer.
-            if state.is_completed {
-                return http_body::SizeHint::with_exact(rem);
-            }
-
-            // Otherwise, the size is the size of the current buffer plus the
-            // size hint returned by the inner body.
-            let (rest_lower, rest_upper) = state
-                .rest
-                .as_ref()
-                .map(|rest| {
-                    let hint = rest.size_hint();
-                    (hint.lower(), hint.upper().unwrap_or(0))
-                })
-                .unwrap_or_default();
-            hint.set_lower(rem + rest_lower);
-            hint.set_upper(rem + rest_upper);
+    #[inline]
+    fn size_hint(&self) -> SizeHint {
+        // If this clone isn't holding the body, return the original size hint.
+        let state = match self.state.as_ref() {
+            Some(state) => state,
+            None => return self.shared.orig_size_hint.clone(),
+        };
+
+        // Otherwise, if we're holding the state but have dropped the inner body, the entire body is
+        // buffered so we know the exact size hint.
+        let buffered = state.buf.remaining() as u64;
+        let rest_hint = match state.rest.as_ref() {
+            Some(rest) => rest.size_hint(),
+            None => return SizeHint::with_exact(buffered),
+        };
+
+        // Otherwise, add the inner body's size hint to the amount of buffered data. An upper limit
+        // is only set if the inner body has an upper limit.
+        let mut hint = SizeHint::default();
+        hint.set_lower(buffered + rest_hint.lower());
+        if let Some(rest_upper) = rest_hint.upper() {
+            hint.set_upper(buffered + rest_upper);
         }
```

Comment:
hmm, if `rest` _doesn't_ have an upper bound, can't we just use the max buffered length as the upper bound, instead of not returning one?

Reply:
But we don't know that it's actually the upper limit of the body, do we? My understanding is that `ReplayBody` should gracefully handle the case when the inner body exceeds the buffer capacity?

Reply:
ah, yeah, it's only the upper limit _if_ we're a subsequent clone of the body --- if we're the initial body, we will continue reading body chunks if we go past the limit; we'll just stop _buffering_ them. we probably _could_ add a check here and only return the max length if we didn't get an upper bound from `rest_hint` *and* we know we're not the original body...but, this is probably fine as-is and it may not be worth bothering with that...